### PR TITLE
Add generated ProtoJSON marshalers

### PR DESCRIPTION
**Pull request by htdvisser**
#### Summary

This pull request is part of https://github.com/TheThingsNetwork/lorawan-stack/issues/2798. It adds JSON marshalers and unmarshalers for our protos, generated by https://github.com/TheThingsIndustries/protoc-gen-go-json.

Actual usage is still disabled by default, but can be enabled by settting the `TTN_LW_EXP_JSONPLUGIN` environment variable to true.

#### Changes

- Add dependency `github.com/TheThingsIndustries/protoc-gen-go-json` v1.1.0
- Generate JSON marshalers with `ghcr.io/thethingsindustries/protoc:3.9.1-gen-go-json-1.1.0`
	- Note: This uses an old version of `protoc` because of our Gogo usage
- Add `(thethings.json.message) = { wrapper: true }` options to all wrapper types and regenerate
- Add `(thethings.json.enum)` and `(thethings.json.enum_value)` options to all enums
- Implement `MarshalProtoJSON` and `UnmarshalProtoJSON` for all types used in `(gogoproto.customtype)`.

#### Testing

- Updated `pkg/ttnpb/ttnpb_encoding_test.go`
- 🐒 with the Console

##### Regressions

Mistakes in `protoc-gen-go-json` or in the options in our `.proto` files could break compatibility of our JSON API. Therefore this is still behind a feature flag, so that we can test this on non-production deployments before fully making the switch.

#### Notes for Reviewers

https://github.com/TheThingsIndustries/protoc-gen-go-json was (or is being) reviewed by @adriansmares and @KrishnaIyer.

I've put this in the v3.15.0 milestone, but if there isn't enough time to review this, we can also postpone it to v3.15.1.

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
	- Since this is an experimental feature behind a feature flag, I don't think we should mention it.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: api/devicerepository.proto [hunk 1]
```diff
@@ -58,25 +59,29 @@ message EndDeviceBrand {
...
+
   // Unknown Key Provisioning.
-  KEY_PROVISIONING_UNKNOWN = 0;
+  KEY_PROVISIONING_UNKNOWN = 0 [(thethings.json.enum_value) = { value: "unknown" }];
   // Custom Key Provisioning.
-  KEY_PROVISIONING_CUSTOM = 1;
+  KEY_PROVISIONING_CUSTOM = 1 [(thethings.json.enum_value) = { value: "custom" }];
   // Key Provisioning from the Global Join Server.
-  KEY_PROVISIONING_JOIN_SERVER = 2;
+  KEY_PROVISIONING_JOIN_SERVER = 2 [(thethings.json.enum_value) = { value: "join server" }];
```
Comment (adriansmares) [inline api/devicerepository.proto:69]: Note to self: Yes, we did have spaces in the textual marshallers.

File: api/third_party/github.com/TheThingsIndustries/protoc-gen-go-json/annotations.proto [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+syntax = "proto2";
+
+package thethings.json;
+
+option go_package = "github.com/TheThingsIndustries/protoc-gen-go-json/annotations";
+
+import "google/protobuf/descriptor.proto";
+
+// TODO: The extension number 1885 is temporary.
+// It will change after we register an extension number on
```
Comment (adriansmares) [inline api/third_party/github.com/TheThingsIndustries/protoc-gen-go-json/annotations.proto]: Is there an issue for this ?

File: pkg/jsonpb/ttn.go [hunk 3]
```diff
@@ -14,25 +14,124 @@
...
+}
+
+type TTNMarshaler struct {
+	*GoGoJSONPb
+}
+
+func (*TTNMarshaler) ContentType() string { return "application/json" }
+
+func (m *TTNMarshaler) Marshal(v interface{}) ([]byte, error) {
+	if marshaler, ok := v.(jsonplugin.Marshaler); ok && jsonpluginFeatureFlag {
```
Comment (adriansmares) [inline pkg/jsonpb/ttn.go]: Nit: I would wrap everything with the feature flag, to avoid the cast completely. Same bellow.

**Comment by adriansmares** [comment]
On a second thought: what if we have another feature flag that allows us to randomly (let's say in 1% of the cases) uses both marshallers and logs an error if the values differ ?

The diff could be obtained by unmarshalling with the `encoding/json` or the `jsoniter` package to `map[string]interface{}`.

File: pkg/types/devaddr.go [hunk 4]
```diff
@@ -71,6 +72,29 @@ func (addr *DevAddr) UnmarshalJSON(data []byte) error {
 	return nil
 }
 
+// MarshalProtoJSON implements the jsonplugin.Marshaler interface.
+func (addr *DevAddr) MarshalProtoJSON(s *jsonplugin.MarshalState) {
+	if addr == nil {
+		s.WriteNil()
```
Comment (johanstokking) [inline pkg/types/devaddr.go:78]: return?

Comment (johanstokking) [reply pkg/types/devaddr.go:78]: Many of these occurrences

**Review by johanstokking — COMMENTED** [review]
Looks very elegant, I had expected much worse. Great stuff!

File: api/third_party/github.com/TheThingsIndustries/protoc-gen-go-json/annotations.proto (hunk 2 above)
Comment (htdvisser) [reply api/third_party/github.com/TheThingsIndustries/protoc-gen-go-json/annotations.proto]: I've changed the number to 51885, which is in the range of numbers allocated for internal use in individual organizations. That way we don't need to register it.

Using 51885 doesn't prevent other projects from using it, as long as it doesn't clash with other extensions.

**Comment by htdvisser** [comment]
@adriansmares @johanstokking I've addressed your comments. Please take another look.

I'd like to get this in for v3.15.1 so that we can start testing it (using the feature flag).

**Review by adriansmares — APPROVED** [review]

**Review by johanstokking — APPROVED** [review]

**Pull request merged**